CRAYSAT-2039: Fix the TestNid2Xname.test_xname2nid_format_range

annapoorna-s-alt opened the pull request:
## Summary and Scope

_Summarize what has changed. Explain why this PR is necessary. What is impacted? Is this a new feature, critical bug fix, etc?_

_Fix the invalid logic of `TestNid2Xname.test_xname2nid_format_range` by removing the wrong assumptions of nid range_

## Issues and Related PRs

_List and characterize relationship to Jira/Github issues and other pull requests. Be sure to list dependencies._

* Resolves [CRAYSAT-2039](https://jira-pro.it.hpe.com:8443/browse/CRAYSAT-2039)

## Testing

_List the environments in which these changes were tested._

### Tested on:

  * rocket
  * tyr

### Test description:

_How were the changes tested and success verified? If schema changes were part of this change, how were those handled in your upgrade/downgrade testing?_

Tested `nid2xname` test cases to ensure it rightly conert nids to xname and vice versa.

## Risks and Mitigations

_Are there known issues with these changes? Any other special considerations?_
NO


## Pull Request Checklist

- [x] Version number(s) incremented, if applicable
- [x] Copyrights updated
- [x] License file intact
- [x] Target branch correct
- [x] CHANGELOG.md updated
- [x] Testing is appropriate and complete, if applicable
- [x] [HPC Product Announcement](https://cray.slack.com/archives/C026TVCSXLH) prepared, if applicable

annapoorna-s-alt commented:
```
ncn-m001:/mnt/developer/sann/SAT/nid2xname/22-9 # ./opt/cray/tests/install/ncn/scripts/python/sat_functional run nid2xname.test_nid2xname
test_multiple_nid_to_xname (csm_testing.tests.sat_functional.nid2xname.test_nid2xname.TestNid2Xname)
Test converting multiple nids to xnames. ... ok
test_multiple_xname_to_nid (csm_testing.tests.sat_functional.nid2xname.test_nid2xname.TestNid2Xname)
Test converting multiple xnames to nids. ... ok
test_nid_range_to_xname (csm_testing.tests.sat_functional.nid2xname.test_nid2xname.TestNid2Xname)
Test converting nid range to xnames using consecutive nids. ... ok
test_nid_to_xname (csm_testing.tests.sat_functional.nid2xname.test_nid2xname.TestNid2Xname)
Test converting nid to xname. ... ok
test_xname2nid_format_nid (csm_testing.tests.sat_functional.nid2xname.test_nid2xname.TestNid2Xname)
Test xname2nid with -f nid option. ... ok
test_xname2nid_format_range (csm_testing.tests.sat_functional.nid2xname.test_nid2xname.TestNid2Xname)
Test xname2nid with -f range option using consecutive nids. ... ok
test_xname_to_nid (csm_testing.tests.sat_functional.nid2xname.test_nid2xname.TestNid2Xname)
Test converting xname to nid. ... ok

----------------------------------------------------------------------
Ran 7 tests in 63.267s

OK

```

haasken-hpe commented on src/csm_testing/tests/sat_functional/nid2xname/test_nid2xname.py:
```diff
@@ -35,6 +35,22 @@ def run_command(self, command):
+        """
+        command = (
+            "cray hsm state components list --type Node --format json | "
+            "jq -r '.Components | map({NID: .NID, Xname: .ID}) | unique | sort_by(.NID)'"
```
Out of curiosity, did you run into an issue that made you add the "unique" filter in your `jq` command? I would think the list of nid, xname pairs would already be unique as both of those should already be unique in HSM.

annapoorna-s-alt replied:
No, I didn't ran into any duplicate issue of xnames but thought of adding unique just for nids because while doing implementation there was an issue with padding zeros left or right. I think we can remove now. Thanks!

haasken-hpe reviewed (approved):
The changes here look good as far as fixing the test failure we've seen in the CT pipeline.

I haven't reviewed these tests before, and I've noticed some other areas for improvement for you to consider:

- Every test method does its own call to "cray hsm state components list". This could be done once in a `setUpClass` method, which would probably make these tests run a bit faster, especially on a larger system.
- The test methods all use `jq` to process the results from the `cray hsm` commands. This requires that we use `shell=True` on the `subprocess.run` calls, which is a bit less secure. The type of processing done here could be done easily in Python, which would have the added benefit of making the tests more comprehensible to the typical Python programmer.
- Some tests (`test_nid_to_xname` and `test_xname_to_nid`) make two calls to `cray hsm state components list` and then rely on the components being ordered in the same way to get the xname and NID of the components. This could be done in a single call like the other tests do.
- The tests `test_multiple_nid_to_xname` and `test_multiple_xname_to_nid` do not seem to actually test anything new. They just run `sat nid2xname` and `sat xname2nid` multiple times for each nid. I think a more useful test would be to call `sat nid2xname` and `sat xname2nid` with multiple nids or xnames given as arguments in a single call. It looks like this is done already for `xname2nid` in `test_xname2nid_format_range` and `test_xname2nid_format_nid`, but it is not done for `nid2xname`.

I'm not sure if it's worth addressing any of these at the moment if the tests are working fine as-is.

annapoorna-s-alt commented:
After addressing the comment, the time reduced to half almost:

```
ncn-m001:/mnt/developer/sann/automation-test/nid2xname/23-9/fin # ./opt/cray/tests/install/ncn/scripts/python/sat_functional run nid2xname.test_nid2xname
test_multiple_nid_to_xname (csm_testing.tests.sat_functional.nid2xname.test_nid2xname.TestNid2Xname)
Test converting multiple nids to xnames in a single command call. ... ok
test_multiple_xname_to_nid (csm_testing.tests.sat_functional.nid2xname.test_nid2xname.TestNid2Xname)
Test converting multiple xnames to nids. ... ok
test_nid_range_to_xname (csm_testing.tests.sat_functional.nid2xname.test_nid2xname.TestNid2Xname)
Test converting nid range to xnames using consecutive nids. ... ok
test_nid_to_xname (csm_testing.tests.sat_functional.nid2xname.test_nid2xname.TestNid2Xname)
Test converting nid to xname. ... ok
test_xname2nid_format_nid (csm_testing.tests.sat_functional.nid2xname.test_nid2xname.TestNid2Xname)
Test xname2nid with -f nid option. ... ok
test_xname2nid_format_range (csm_testing.tests.sat_functional.nid2xname.test_nid2xname.TestNid2Xname)
Test xname2nid with -f range option using consecutive nids. ... ok
test_xname_to_nid (csm_testing.tests.sat_functional.nid2xname.test_nid2xname.TestNid2Xname)
Test converting xname to nid. ... ok

----------------------------------------------------------------------
Ran 7 tests in 26.647s

OK

```

haasken-hpe commented on src/csm_testing/tests/sat_functional/nid2xname/test_nid2xname.py:
```diff
@@ -30,140 +30,144 @@
+        """Get component data from HSM."""
+        # Get raw HSM data - no jq processing needed
+        command = "cray hsm state components list --type Node --format json"
+        result = subprocess.run(command, shell=True, check=True, stdout=subprocess.PIPE, stderr=subprocess.PIPE)
```
I'm pretty sure you don't need `shell=True` here anymore since you're not piping into `jq`.

```suggestion
        result = subprocess.run(command, check=True, stdout=subprocess.PIPE, stderr=subprocess.PIPE)
```

haasken-hpe commented on src/csm_testing/tests/sat_functional/nid2xname/test_nid2xname.py:
```diff
@@ -30,140 +30,144 @@
+    @classmethod
+    def get_hsm_components(cls):
+        """Get component data from HSM."""
+        # Get raw HSM data - no jq processing needed
```
Some of these comments look like artifacts of asking GitHub Copilot to make these changes. I would delete them, since they don't add much, IMO. They only really make sense when looked at compared to the earlier version of the code.

```suggestion
```

haasken-hpe commented on src/csm_testing/tests/sat_functional/nid2xname/test_nid2xname.py:
```diff
@@ -30,140 +30,144 @@
+        """Set up data used by all test methods."""
+        super().setUpClass()
+        cls.components = cls.get_hsm_components()
+        # Pre-calculate the first node's data for simple tests
```
Also not a super important comment, IMO:

```suggestion
```

haasken-hpe commented on src/csm_testing/tests/sat_functional/nid2xname/test_nid2xname.py:
```diff
@@ -30,140 +30,144 @@
+        command = "cray hsm state components list --type Node --format json"
+        result = subprocess.run(command, shell=True, check=True, stdout=subprocess.PIPE, stderr=subprocess.PIPE)
+
+        # Process the data in Python instead of jq
```
Another comment that mostly only makes sense when we compare to the previous version of the code.


```suggestion
```

haasken-hpe commented on src/csm_testing/tests/sat_functional/nid2xname/test_nid2xname.py:
```diff
@@ -30,140 +30,144 @@
+
     def run_command(self, command):
         """Run a shell command and return the output."""
         result = subprocess.run(command, shell=True, check=True, stdout=subprocess.PIPE, stderr=subprocess.PIPE)
```
I don't believe there is any need for `shell=True` here anymore.


```suggestion
        result = subprocess.run(command, check=True, stdout=subprocess.PIPE, stderr=subprocess.PIPE)
```

haasken-hpe commented on src/csm_testing/tests/sat_functional/nid2xname/test_nid2xname.py:
```diff
@@ -30,140 +30,144 @@
+        self.assertEqual(expected_xnames, converted_xnames,
+                        f"Comma-separated test: Expected xnames {expected_xnames} but got {converted_xnames}")
 
     def test_multiple_xname_to_nid(self):
```
I don't think this method is testing anything not already tested by the other test methods. It is basically just repeating the test from `test_xname_to_nid` three times with the first three xnames. I suppose it doesn't hurt anything though. The `test_xname2nid_format_range` and `test_xname2nid_format_nid` tests already test `xname2nid` when given multiple xnames in a single command.

annapoorna-s-alt replied:
I have addressed to get output as nid ranges by default for multiple xnames.
